Add a DogStatsd sink

talwai opened the pull request:
This adds a very basic sink for [DogStatsd](http://docs.datadoghq.com/guides/dogstatsd/) implementing the MetricSink
interface.
It has two main features:
1. Using the client at https://github.com/DataDog/datadog-go/statsd it
allows global tags to be assigned to all metrics sent into the sink
2. When necessary, it splices out the hostname from incoming metric
names, and optinally delegates host tagging to the Dogstatsd server. This allows
for much saner grouping by host in the Datadog backend.

talwai commented:
@armon - ping. Any thoughts on this?

armon commented:
@talwai Could you move the sink into a sub-package called `datadog`? That way you do not add the data dog statsd client lib as a dependency if not used.

armon commented on dogstatsd.go:
```diff
@@ -0,0 +1,115 @@
+)
+
+// HostnameGetter is a generic function to retrieve a hostname
+type HostnameGetter func() string
```
Why is this exported?

armon commented on dogstatsd.go:
```diff
@@ -0,0 +1,115 @@
+// with a dogstatsd server. It utilizes the Dogstatsd client at github.com/Datadog/datadog-go/statsd
+type DogStatsdSink struct {
+	client            *statsd.Client
+	hostnameGetter    HostnameGetter
```
Can the hostname just be a static value? Seems odd for it to change

armon commented on dogstatsd.go:
```diff
@@ -0,0 +1,115 @@
+}
+
+// NewDogStatsdSink is used to create a new DogStatsdSink with sane defaults
+func NewDogStatsdSink(addr string) (*DogStatsdSink, error) {
```
Just pass in the hostname here, or use os.Hostname()

armon commented on dogstatsd.go:
```diff
@@ -0,0 +1,115 @@
+	propagateHostname bool
+}
+
+func getHostname() string {
```
Not needed most likely

armon commented on dogstatsd.go:
```diff
@@ -0,0 +1,115 @@
+	return sink, nil
+}
+
+func (s *DogStatsdSink) setTags(tags []string) {
```
What is this function for?

armon commented on dogstatsd.go:
```diff
@@ -0,0 +1,115 @@
+	s.client.Tags = tags
+}
+
+func (s *DogStatsdSink) enableHostnamePropagation() {
```
Similarly, this function seems weird

armon commented on start.go:
```diff
@@ -73,6 +74,14 @@ func NewGlobal(conf *Config, sink MetricSink) (*Metrics, error) {
 }
 
+// GetConfig returns the Config for the globalMetrics instance if available
+func GetConfig() (*Config, error) {
```
I think it should just return *Config which may be nil, no need for an error return.

talwai commented:
@armon Sorry for the delayed response. I think I've addressed the majority of your issues in code and added more documentation around the rest to explain my approach. Let me know how this looks. Also I moved the sink into a `datadog` subpackage as you suggested

armon commented:
Thanks!
